api/main.py

```diff
@@ -186,6 +188,48 @@ async def get_root_node(node_id: str):
     return node
 
 
+@app.get('/trigger_completed_event/{node_id}')
+async def trigger_completed_event(node_id: str, wait_time_hours: int = 0,
+                                  wait_time_minutes: int = 0,
+                                  wait_time_seconds: int = 0):
+    """Trigger an event when all child nodes are completed of
+       a given node"""
+    try:
+        nodes = await db.find_by_attributes(Node,
+                                            {"parent": ObjectId(node_id)})
+
+        if len(nodes) == 0:
+            raise HTTPException(
+                status_code=status.HTTP_400_BAD_REQUEST,
+                detail=f"No child node found of a given node id:{node_id}"
+            )
+
+        for node in nodes:
+            if node.status == 'pending':
+                current_time = datetime.utcnow()
+                timeout = node.created + timedelta(hours=wait_time_hours,
+                                                   minutes=wait_time_minutes,
+                                                   seconds=wait_time_seconds)
+                await Node.wait_for_node(current_time, timeout)
+
+                ret = node.set_timeout_status(current_time)
+                if ret:
+                    await db.update(node)
+                else:
+                    return {"message": "Nodes are not completed yet"}
+
+        operation = 'completed'
+        await pubsub.publish_cloudevent('node', {'op': operation,
+                                                 'id': str(node_id),
+                                        'nodes': jsonable_encoder(nodes)})
+        return {"message": "'Event triggered"}
```

Comment:
Is it possible to narrow down the block which is in the `try..except` block? I'm not entirely sure that we need to keep the whole function body in it.

Reply:
Yes, I narrowed it down.